doc/flow_counters/routes_flow_counters.md

```diff
@@ -0,0 +1,476 @@
+# SONiC Route Flow Counter Design #
+
+## Table of Content
+
+### Revision
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Junchao Chen  | Initial version                   |
+
+### Scope
+
+This document is the design document for route flow counter feature on SONiC.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+Flow counters are usually used for debugging, troubleshooting and performance enhancement processes. Flow counters could cover cases like:
+
+- Routes matching the configured prefix pattern (number of hits and number of bytes)
+- Host interface traps (number of received traps per Trap ID)
+- FDB entries matching the configured VXLAN tunnel or using the VLAN ID as pattern
+- Next-Hop/Next-Hop Group/Next-Hop Group Member
+
+This document focus on route counter.
+
+### Requirements
+
+- Generic Counters shall be used as Flow Counters introduced by the feature
+- CLI shall be used for configuration, showing and clearing of statistics
+- Flow Counters for routes shall be configured using prefix patterns. The flow counter shall be bound to all routes matching the configured pattern (prefix:vrf)
+- In Phase 1 the number of configured route patterns shall be limited to 1 (enforcement shall be done during configuration via CLI)
+- In Phase 1 the number of matching routes shall be limited to the pre-configured value (default value - 30), after reboot it is not ensured that the same set of matching routes will be used for counting
+- Flow Counters shall be bound the matching routes regardless how these routes are added - manually (static) or via FRR
+- Statistics shall be configured (enabled/disabled)  and cleared using the CLI commands
+- Statistics shall be provided as a number of hit/use of a specific resource and number of bytes in packets sent via configured routes
+- When a prefix pattern is removed, Flow Counters on all routes matching the configured pattern shall be removed
+- Adding route entry shall be automatically bound to counter if counter is enabled and pattern matches
+- Removing route entry shall be automatically unbound if the entry is previously bound
+- Removing route entry shall not trigger route pattern search
+- Decreasing max allowed match count shall not trigger removing of existing bound counters
```

Comment:
Does this mean after max allowed the actual number bounded could be more than the max number allowed>

Reply:
can keep the inactive set for the pattern in the cache, and if the active set has vaccancy, then you can add the flow in the inactive set back into the active set?